modules/generic/src/main/scala/pureconfig/generic/EnumerationConfigReaderBuilder.scala

```diff
@@ -31,11 +37,12 @@ object EnumerationConfigReaderBuilder {
     hGen: LabelledGeneric.Aux[H, HNil],
     tReaderBuilder: EnumerationConfigReaderBuilder[T]): EnumerationConfigReaderBuilder[FieldType[K, H] :+: T] =
     new EnumerationConfigReaderBuilder[FieldType[K, H] :+: T] {
-      def build(transformName: String => String): ConfigReader[FieldType[K, H] :+: T] = {
-        lazy val tReader = tReaderBuilder.build(transformName)
+      def build(transformName: String => String, validOptions: Set[String]): ConfigReader[FieldType[K, H] :+: T] = {
+        val transformedName = transformName(vName.value.name)
+        lazy val tReader = tReaderBuilder.build(transformName, validOptions + transformedName)
```

Comment:
I think this would be slightly easier to reason about if we wouldn't mix a bottom-up approach (creating a reader for the remaining coproduct alternatives) with a top-down one (supplying the options previously tried).

Do you know if there's a native shapeless way to extract the names of the types from the `Coproduct` representation `LabelledGeneric` produces?

Reply:
@jcazevedo It looks like there's a pretty easy way to get a [list of the names](https://stackoverflow.com/a/29679342), but I don't see way to get a similar way to get a list of the instances without using a ["bottom-up" approach of destructing the `LabelleledGeneric`](https://stackoverflow.com/a/43658710). 

If you can find a way to simply get a list of the instances, I'd love to do it that way instead.

Reply:
The solution that plays well with the structurally recursive nature of shapeless generics is to get rid of this parameter, start by failing with an empty set of options in `EnumerationConfigReaderBuilder[CNil]` and keep prepending options to the failure in the recursive case (i.e. changing line 45 to something like `tReader.from(cur).right.map(Inr.apply).left.map(maybeAddOptionToKeyNotFound)`. We'd need some extra boilerplate for destructuring and reconstructing the failure when we are going "back up" (probably a single pattern match is enough).

Reply:
@ruippeixotog I struggle with shapeless enough that what you're describing isn't obvious to me. Can you think of an example similar to what you described that I could refer to?

Reply:
Unfortunately I don't think we have similar code examples. What I'm suggesting is to get rid of that parameter and follow a bottom-up, inductive approach:

- **Base case**: the coproduct is empty (e.g. an empty sealed trait). The result is simply `cur.failed(NoValidEnumOptionFound(cur.value, Set.empty, enumType))`.

- **Inductive case**: we're dealing with a coproduct of type `FieldType[K, H] :+: T`. We do just like we are doing now, but we map both successful results (wrapping it in `Inr`) and failed results with type `NoValidEnumOptionFound` to add the attempted option. In code this would be replacing line 41 with something like:

```scala
case Right(_) =>
  tReader.from(cur) match {
    case Right(res) => Right(Inr(res))
    case Left(ConfigReaderFailures(NoValidEnumOptionFound(value, validOptions, enumType))) =>
      Left(ConfigReaderFailures(NoValidEnumOptionFound(value, validOptions + transformedName, enumType)))
    case Left(failures) => Left(failures)
  }.right.map(Inr.apply)
```

I didn't compile this, it can probably be simplified and some care has to be taken with pattern matching on `ConfigReaderFailures` like this, but I hope it helps explaining what I meant :)